### DISCUSSION TITLE
Normalize the hue of all images at once from a reference image

### DISCUSSION THREAD

**Discussion by flexeo** [ImageMagick/ImageMagick#7803] [Help] [answered]
### ImageMagick version

7

### Operating system, version and so on

Windows 10

### Description

I've got several images of characters from A-Z and 0-9 that have been generated with an artificial intelligence program and the problem is that all these characters don't have exactly the same hue because they've been generated one by one, one after the other. How can I perform batch processing to standardize the hue of all the images at once, taking one of the images in the batch as the hue reference?
![alphanum](https://github.com/user-attachments/assets/93da1aff-6a80-4d20-a497-3ada24a7271a)

**Comment by snibgo** [comment]
Please post two of the PNG files.

Perhaps they are coloured characters on transparent background. If so, then the average hue might be calculated by (Windows syntax):
```
magick 0.png -colorspace HCL -channel 0 -separate -scale "1x1^!" -define ftxt:format=\p\n ftxt:
```
(EDIT: This converts every pixel to Hue, and finds the average of those. Perhaps it would be better to find the average of the RGB values, then convert that to Hue.)

The text output will be something like:
```
32.3453%
```
This is the average hue of 0.png. Set an environment variable, such as REQHUE, to this value. We can then set the hue of any other image:
```
magick 1.png -colorspace HCL -channel 0 -evaluate set %REQHUE% +channel -colorspace sRGB out_1.png
```

**Reply by flexeo** [reply]
![0](https://github.com/user-attachments/assets/3e27b740-589e-4435-b6f2-94c7cc544ccc)
I don't get the same value at all with the latest version of Imagemagick on my system (Windows):

`magick 0.png -colorspace HCL -channel 0 -separate -scale "1x1^!" -define ftxt:format=\p\n ftxt:`

The text output will be something like:
`13.5978%`

If I test with the value returned to me, I get this:
`magick 1.png -colorspace HCL -channel 0 -evaluate set "13.5978%" +channel -colorspace sRGB out_1.png`

![out_1](https://github.com/user-attachments/assets/aa98fac4-14b5-4d4b-8dde-5e8657802cb1)

Why don't I get the value "32.3453%" like you? Because your final result is perfect :
![out_1](https://github.com/user-attachments/assets/91640ece-2b85-4ea2-9707-90058e828cb2)

**Comment by snibgo** [comment, marked as answer]
My first command was wrong. The problem is that "-colorspace HCL -channel 0 -separate" removes the transparency, so the "-scale" takes the average of all the pixels, including the pixels that were transparent.

My second thought gives the correct result:
```
magick 0.png -scale "1x1^!" -colorspace HCL -channel 0 -separate -define ftxt:format=\p\n ftxt:
31.7587%
```

> Why don't I get the value "32.3453%" like you? Because your final result is perfect 

I didn't have your images, so I used my faulty command on another (opaque) image. By coincidence, it gave roughly the number we need.

**Comment by flexeo** [comment]
Thank you very much for your clear answer.

The command works very well when the batch of images is mainly made up of a single color, but I have batches of images that include several colors with variations in brightness/contrast/saturation between them and inevitably, it doesn't work.

Is there a way to standardize a batch of images from a reference image in this case?

Here's an example:
![multicolor](https://github.com/user-attachments/assets/ccdadfb4-5695-4e5e-82dd-a4fab9785ed8)
![0](https://github.com/user-attachments/assets/979d9abf-b92d-45c9-841d-6f81da2164c6)
![u](https://github.com/user-attachments/assets/c367e9c7-1770-4e69-9bd5-f04e7e33dc49)

**Comment by snibgo** [comment]
In the single colour problem, I used a simple rule to transform the images: calculate the average hue of the reference image, and change each other image to that hue.

In the multiple colours problem, I can't see what simple rule to apply. Or even what complex rule to apply. For example, at the colour boundaries, 0.png has blurred blending but U.png has sharper boundaries more like watercolour painting.

**Comment by flexeo** [comment]
Thank you very much for your help.
